File: src/interactive.py
#!/usr/bin/env python3
"""Interactive mode for Thermalright LCD Animations with keyboard controls."""
import sys
import time
import random
import tty
import termios
import select
from lcd_controller import LCDController
from animation_library import ANIMATIONS


class InteractiveController:
    """Interactive animation controller with keyboard navigation."""

    def __init__(self, lcd, config):
        self.lcd = lcd
        self.config = config
        self.animation_names = list(ANIMATIONS.keys())
        self.current_index = 0
        self.current_anim = None
        self.duration = 10.0  # Default duration in seconds
        self.update_interval = config.get('update_interval', 0.005)
        self.random_mode = True
        self.start_time = time.time()

    def get_key(self):
        """Non-blocking keyboard input."""
        if select.select([sys.stdin], [], [], 0)[0]:
            return sys.stdin.read(1)
        return None

    def draw_ui(self):
        """Draw the status UI."""
        print("\033[2J\033[H", end='')  # Clear screen
        print("=" * 70)
        print(" THERMALRIGHT LCD ANIMATIONS - Interactive Mode")
        print(" (TyphooN@marketwizardry.org)")
        print("=" * 70)
        print()
        print(f" Current Animation: [{self.current_index + 1}/{len(self.animation_names)}]")
        print(f" {self.animation_names[self.current_index]}")
        print()
        print(f" Mode: {'RANDOM' if self.random_mode else 'MANUAL'}")
        print(f" Duration: {self.duration:.1f}s")
        print(f" Speed (interval): {self.update_interval:.3f}s")
        print()
        print("-" * 70)
        print(" Controls:")
        print("   ← →  : Previous/Next animation")
        print("   SPACE: Jump to random animation")
        print("   m    : Toggle Manual/Random mode")
        print("   + -  : Increase/Decrease duration (±1s)")
        print("   [ ]  : Decrease/Increase speed")
        print("   q    : Quit")
        print("-" * 70)
        print()

    def next_animation(self):
        """Go to next animation."""
        self.current_index = (self.current_index + 1) % len(self.animation_names)
        self.load_current_animation()

    def prev_animation(self):
        """Go to previous animation."""
        self.current_index = (self.current_index - 1) % len(self.animation_names)
        self.load_current_animation()

    def random_animation(self):
        """Select random animation."""
        self.current_index = random.randint(0, len(self.animation_names) - 1)
        self.load_current_animation()

    def load_current_animation(self):
        """Load the current animation."""
        anim_name = self.animation_names[self.current_index]
        anim_class = ANIMATIONS[anim_name]
        self.current_anim = anim_class(self.lcd)
        self.current_anim.reset()
        self.start_time = time.time()
        self.draw_ui()

    def adjust_duration(self, delta):
        """Adjust duration by delta seconds."""
        self.duration = max(1.0, self.duration + delta)
        self.draw_ui()

    def adjust_speed(self, faster=True):
        """Adjust animation speed."""
        if faster:
            self.update_interval = max(0.001, self.update_interval - 0.005)
        else:
            self.update_interval = min(0.2, self.update_interval + 0.005)
        self.draw_ui()

    def toggle_random(self):
        """Toggle random mode."""
        self.random_mode = not self.random_mode
        self.draw_ui()

    def run(self):
        """Run the interactive mode."""
        # Set up terminal for raw input
        old_settings = termios.tcgetattr(sys.stdin)
        try:
            tty.setcbreak(sys.stdin.fileno())

            # Load first animation
            self.load_current_animation()

            while True:
                # Check for keyboard input
                key = self.get_key()

                if key:
                    if key == 'q':
                        break
                    elif key == '\x1b':  # Arrow keys start with escape
                        # Read the rest of the arrow key sequence
                        next1 = sys.stdin.read(1)
                        next2 = sys.stdin.read(1)
                        if next1 == '[':
                            if next2 == 'C':  # Right arrow
                                self.next_animation()
                            elif next2 == 'D':  # Left arrow
                                self.prev_animation()
                    elif key == ' ':  # Spacebar
                        self.random_animation()
                    elif key == 'm' or key == 'M':  # Toggle mode
                        self.toggle_random()
                    elif key == '+' or key == '=':
                        self.adjust_duration(1.0)
                    elif key == '-' or key == '_':
                        self.adjust_duration(-1.0)
                    elif key == '[':
                        self.adjust_speed(faster=False)
                    elif key == ']':
                        self.adjust_speed(faster=True)

                # Check if duration expired in random mode
                if self.random_mode:
                    if time.time() - self.start_time >= self.duration:
                        self.random_animation()

                # Update animation
                if self.current_anim:
                    self.current_anim.update()
                    self.lcd.send_packets()

                time.sleep(self.update_interval)

        finally:
            # Restore terminal settings
            termios.tcsetattr(sys.stdin, termios.TCSADRAIN, old_settings)
            # Clear screen and show cursor
            print("\033[2J\033[H", end='')
            print("\nExiting interactive mode...")


def main():
    """Main entry point for interactive mode."""
    import json
    import os

    # Load config
    config_path = os.path.join(
        os.path.dirname(os.path.dirname(__file__)),
        'config.json'
    )

    try:
        with open(config_path, 'r') as f:
            config = json.load(f)
    except Exception as e:
        print(f"Error loading config: {e}")
        config = {}

    # Initialize LCD
    vendor_id = int(config.get('vendor_id', '0x0416'), 16)
    product_id = int(config.get('product_id', '0x8001'), 16)

    print("\nInitializing LCD Controller...")
    lcd = LCDController(vendor_id=vendor_id, product_id=product_id)

    if lcd.dev is None:
        print("\nError: Could not connect to LCD device.")
        return 1

    print("Connected successfully!\n")
    print("Starting interactive mode in 2 seconds...")
    time.sleep(2)

    # Run interactive controller
    controller = InteractiveController(lcd, config)
    try:
        controller.run()
    finally:
        lcd.clear()
        lcd.send_packets()
        lcd.close()

    return 0


if __name__ == '__main__':
    sys.exit(main())

File: src/interactive.rs
use std::io::{stdout, Write};
use std::time::{Duration, Instant};

use crossterm::cursor::{Hide, MoveTo, Show};
use crossterm::event::{self, Event, KeyCode, KeyModifiers};
use crossterm::style::Print;
use crossterm::terminal::{
    disable_raw_mode, enable_raw_mode, Clear, ClearType, EnterAlternateScreen, LeaveAlternateScreen,
};
use crossterm::{execute, queue};
use rand::Rng;

use crate::animations::{self, Animation};
use crate::lcd::LcdController;

pub struct InteractiveController {
    lcd: LcdController,
    entries: &'static [animations::Entry],
    current_index: usize,
    current_anim: Option<Box<dyn Animation>>,
    duration: f32,
    update_interval: f32,
    random_mode: bool,
    start_time: Instant,
}

impl InteractiveController {
    pub fn new(lcd: LcdController, update_interval: f32) -> Self {
        Self {
            lcd,
            entries: animations::all(),
            current_index: 0,
            current_anim: None,
            duration: 10.0,
            update_interval,
            random_mode: true,
            start_time: Instant::now(),
        }
    }

    fn load_current(&mut self) {
        let entry = &self.entries[self.current_index];
        let mut anim = (entry.factory)();
        anim.reset(&mut self.lcd);
        self.duration = anim.preferred_duration();
        self.current_anim = Some(anim);
        self.start_time = Instant::now();
        let _ = self.draw_ui();
    }

    fn next(&mut self) {
        self.current_index = (self.current_index + 1) % self.entries.len();
        self.load_current();
    }

    fn prev(&mut self) {
        self.current_index = if self.current_index == 0 {
            self.entries.len() - 1
        } else {
            self.current_index - 1
        };
        self.load_current();
    }

    fn random(&mut self) {
        let mut rng = rand::thread_rng();
        self.current_index = rng.gen_range(0..self.entries.len());
        self.load_current();
    }

    fn adjust_duration(&mut self, delta: f32) {
        self.duration = (self.duration + delta).max(1.0);
        let _ = self.draw_ui();
    }

    fn adjust_speed(&mut self, faster: bool) {
        if faster {
            self.update_interval = (self.update_interval - 0.005).max(0.001);
        } else {
            self.update_interval = (self.update_interval + 0.005).min(0.2);
        }
        let _ = self.draw_ui();
    }

    fn toggle_random(&mut self) {
        self.random_mode = !self.random_mode;
        let _ = self.draw_ui();
    }

    fn draw_ui(&self) -> std::io::Result<()> {
        let mut out = stdout();
        let entry = &self.entries[self.current_index];
        queue!(out, Clear(ClearType::All), MoveTo(0, 0))?;
        queue!(
            out,
            Print("======================================================================\r\n")
        )?;
        queue!(
            out,
            Print(" THERMALRIGHT LCD ANIMATIONS - Interactive Mode (Rust)\r\n")
        )?;
        queue!(
            out,
            Print("======================================================================\r\n")
        )?;
        queue!(out, Print("\r\n"))?;
        queue!(
            out,
            Print(format!(
                " Current Animation: [{}/{}]\r\n",
                self.current_index + 1,
                self.entries.len()
            ))
        )?;
        queue!(out, Print(format!(" {} - {}\r\n", entry.name, entry.desc)))?;
        queue!(out, Print("\r\n"))?;
        queue!(
            out,
            Print(format!(
                " Mode: {}\r\n",
                if self.random_mode { "RANDOM" } else { "MANUAL" }
            ))
        )?;
        queue!(
            out,
            Print(format!(
                " Dwell: {:.1}s (animation recommended)\r\n",
                self.duration
            ))
        )?;
        queue!(
            out,
            Print(format!(
                " Speed (interval): {:.3}s\r\n",
                self.update_interval
            ))
        )?;
        queue!(out, Print("\r\n"))?;
        queue!(
            out,
            Print("----------------------------------------------------------------------\r\n")
        )?;
        queue!(out, Print(" Controls:\r\n"))?;
        queue!(out, Print("   <- ->  : Previous/Next animation\r\n"))?;
        queue!(out, Print("   SPACE  : Jump to random animation\r\n"))?;
        queue!(out, Print("   m      : Toggle Manual/Random mode\r\n"))?;
        queue!(
            out,
            Print("   + -    : Increase/Decrease duration (+/- 1s)\r\n")
        )?;
        queue!(out, Print("   [ ]    : Decrease/Increase speed\r\n"))?;
        queue!(out, Print("   q      : Quit\r\n"))?;
        queue!(
            out,
            Print("----------------------------------------------------------------------\r\n")
        )?;
        out.flush()
    }

    pub fn run(&mut self) -> std::io::Result<()> {
        enable_raw_mode()?;
        execute!(stdout(), EnterAlternateScreen, Hide)?;

        if self.random_mode {
            self.random();
        } else {
            self.load_current();
        }

        let result = self.event_loop();

        execute!(stdout(), Show, LeaveAlternateScreen)?;
        disable_raw_mode()?;

        self.lcd.clear();
        self.lcd.send_packets();

        result
    }

    fn event_loop(&mut self) -> std::io::Result<()> {
        loop {
            if event::poll(Duration::from_secs(0))? {
                if let Event::Key(key) = event::read()? {
                    match key.code {
                        KeyCode::Char('q') | KeyCode::Char('Q') => break,
                        KeyCode::Char('c') if key.modifiers.contains(KeyModifiers::CONTROL) => {
                            break
                        }
                        KeyCode::Right => self.next(),
                        KeyCode::Left => self.prev(),
                        KeyCode::Char(' ') => self.random(),
                        KeyCode::Char('m') | KeyCode::Char('M') => self.toggle_random(),
                        KeyCode::Char('+') | KeyCode::Char('=') => self.adjust_duration(1.0),
                        KeyCode::Char('-') | KeyCode::Char('_') => self.adjust_duration(-1.0),
                        KeyCode::Char('[') => self.adjust_speed(false),
                        KeyCode::Char(']') => self.adjust_speed(true),
                        _ => {}
                    }
                }
            }

            if self.random_mode {
                if self.start_time.elapsed() >= Duration::from_secs_f32(self.duration) {
                    self.random();
                }
            }

            if let Some(anim) = self.current_anim.as_mut() {
                anim.update(&mut self.lcd);
                self.lcd.send_packets();
            }

            std::thread::sleep(Duration::from_secs_f32(self.update_interval));
        }
        Ok(())
    }
}
